### ISSUE TITLE
Incorrect result size: intermittent UAA failures

### ISSUE THREAD

**Issue by combor** [cloudfoundry/uaa#372]
## Issue

UAA fails to modify user groups while executing multiple user modify requests in parallel. 

## Context

We were able to reproduce this failure by running two processes that are modifying user groups in parallel . We've got two UAAs and we were running two user group modifications concurrently. Both scripts were modifying different users. The transaction with UAA looks like this:

```
--->
request: put https://uaa.piotr-dev.dev.cloudpipeline.digital/Groups/1c4f1b37-e0c0-4a39-ace8-4ccddc340862
body: {"meta":{"version":1090,"created":"2016-05-23T10:13:37.661Z","lastModified":"2016-05-24T13:33:15.440Z"},"schemas":["urn:scim:schemas:core:1.0"],"members":["bfac0099-94b8-45a9-b33b-a0dbbe521623","af459776-8d16-41fe-bd32-e86b13d3b321"],"id":"1c4f1b37-e0c0-4a39-ace84ccddc340862","displayName":"cloud_controller.admin","zoneid":"uaa"}

<---
response: 409 
body: {"error_description":"Incorrect result size: expected 1, actual 0","error":"scim","message":"Incorrect result size: expected 1, actual 0"}
error response:
{
  "error_description": "Incorrect result size: expected 1, actual 0",
  "error": "scim",
  "message": "Incorrect result size: expected 1, actual 0"
}
```

Versions:

* CF v235
* uaa-release 68cf1f1684
* uaa 616e8bcc58
* cf-uaac gem 3.2.0

## Steps to Reproduce

Run two instances of [user creation script](https://gist.github.com/saliceti/38109ff695347eaec30110317b4e881a) in the loop for instance: ` while [[ $? == 0 ]];do ./admin_user.sh test;done`.

## Expected result

`success` for every `uaa member add`

## Current result

```
{
  "error_description": "Incorrect result size: expected 1, actual 0",
  "error": "scim",
  "message": "Incorrect result size: expected 1, actual 0"
}
```

## SQL queries

These are queries that are executed just before the failure:
```
[2016-05-24 13:33:19.667] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL statement [update groups set version=?, displayName=?, description=?, lastModified=? where id=? and version=? and identity_zone_id=?]
[2016-05-24 13:33:19.668] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: SQL update affected 0 rows
[2016-05-24 13:33:19.669] uaa - 6510 [http-bio-8080-exec-29] .... ERROR --- ScimGroupEndpoints: Error updating group, restoring to previous state
[2016-05-24 13:33:19.669] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL query
[2016-05-24 13:33:19.669] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL statement [select id,displayName,description,created,lastModified,version,identity_zone_id from groups where id=? and identity_zone_id=?]
[2016-05-24 13:33:19.670] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL update
[2016-05-24 13:33:19.670] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL statement [update groups set version=?, displayName=?, description=?, lastModified=? where id=? and version=? and identity_zone_id=?]
[2016-05-24 13:33:19.672] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: SQL update affected 1 rows
[2016-05-24 13:33:19.672] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL query
[2016-05-24 13:33:19.672] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- JdbcTemplate: Executing prepared SQL statement [select id,displayName,description,created,lastModified,version,identity_zone_id from groups where id=? and identity_zone_id=?]
[2016-05-24 13:33:19.673] uaa - 6510 [http-bio-8080-exec-29] .... DEBUG --- ExceptionHandlerExceptionResolver: Resolving exception from handler [public org.cloudfoundry.identity.uaa.scim.ScimGroup org.cloudfoundry.identity.uaa.scim.endpoints.ScimGroupEndpoints.updateGroup(org.cloudfoundry.identity.uaa.scim.ScimGroup,java.lang.String,java.lang.String,javax.servlet.http.HttpServletResponse)]: org.cloudfoundry.identity.uaa.scim.exception.ScimException: Incorrect result size: expected 1, actual
```
## Stack trace

```
org.cloudfoundry.identity.uaa.scim.exception.ScimException: Incorrect result size: expected 1, actual 0
    at org.cloudfoundry.identity.uaa.scim.endpoints.ScimGroupEndpoints.updateGroup(ScimGroupEndpoints.java:372)
    at org.cloudfoundry.identity.uaa.scim.endpoints.ScimGroupEndpoints$$FastClassBySpringCGLIB$$79976b44.invoke(<generated>)
    at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
    at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:717)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
    at org.springframework.aop.framework.adapter.AfterReturningAdviceInterceptor.invoke(AfterReturningAdviceInterceptor.java:52)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653)
    at org.cloudfoundry.identity.uaa.scim.endpoints.ScimGroupEndpoints$$EnhancerBySpringCGLIB$$10f9bd38.updateGroup(<generated>)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
    at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
    at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:110)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:806)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:729)
    at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
    at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:883)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:653)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.cloudfoundry.identity.uaa.web.BackwardsCompatibleScopeParsingFilter.doFilter(BackwardsCompatibleScopeParsingFilter.java:40)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:316)
    at org.cloudfoundry.identity.uaa.authentication.SessionResetFilter.doFilterInternal(SessionResetFilter.java:89)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.security.web.XFrameOptionsFilter.doFilterInternal(XFrameOptionsFilter.java:16)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
    at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:169)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.scim.DisableInternalUserManagementFilter.doFilterInternal(DisableInternalUserManagementFilter.java:56)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.scim.DisableUserManagementSecurityFilter.doFilterInternal(DisableUserManagementSecurityFilter.java:88)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.zone.IdentityZoneSwitchingFilter.doFilterInternal(IdentityZoneSwitchingFilter.java:139)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationProcessingFilter.doFilter(OAuth2AuthenticationProcessingFilter.java:176)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:53)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:91)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.security.web.SecurityFilterChainPostProcessor$UaaLoggingFilter.doFilter(SecurityFilterChainPostProcessor.java:253)
    at org.cloudfoundry.identity.uaa.security.web.SecurityFilterChainPostProcessor$HttpsEnforcementFilter.doFilter(SecurityFilterChainPostProcessor.java:196)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.oauth.DisableIdTokenResponseTypeFilter.doFilterInternal(DisableIdTokenResponseTypeFilter.java:87)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.zone.IdentityZoneResolvingFilter.doFilterInternal(IdentityZoneResolvingFilter.java:66)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.cloudfoundry.identity.uaa.security.web.CorsFilter.doFilterInternal(CorsFilter.java:222)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
    at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:213)
    at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:176)
    at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
    at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
    at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
    at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:683)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:620)
    at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 0
    at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:183)
    at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:48)
    at org.cloudfoundry.identity.uaa.scim.endpoints.ScimGroupEndpoints.updateGroup(ScimGroupEndpoints.java:358)
    ... 106 more
```

**Comment by plfx** [comment]
Hello @combor . These errors will occur whenever you have a client (or thread) trying to update a group with out-of-date data, as tracked by the `version` number in the JSON request you send.

We provide the `/Groups/{groupId}/members` endpoints for performing group membership operations. To add a member to a group, perform a `POST` to this endpoint as specified here: http://docs.cloudfoundry.com/uaa/#add-member

I'm marking this issue closed. Please contact us again if there are any problems with this approach.

**Comment by combor** [comment]
Hi @plfx . Thanks for explanation. The client in this case, as you can see in provided user mod script , is `uuac` and transaction dump is from `uaac --trace`. So I think this issue has been closed prematurely.

I'll paste this script here bot better readability:
```
#!/bin/sh
set -eu
NAME=$1

UAA_ENDPOINT=https://...
UAA_ADMIN_CLIENT_PASS=xxxxx
PASSWORD=test

echo "Creating user ${NAME}"
uaac --trace target "${UAA_ENDPOINT}"
uaac --trace token client get admin -s "${UAA_ADMIN_CLIENT_PASS}"
uaac --trace user add "${NAME}" -p "${PASSWORD}" --emails ignored

echo "Add permissions to user ${NAME}"
uaac --trace member add cloud_controller.admin "${NAME}"
uaac --trace member add uaa.admin "${NAME}"
uaac --trace member add scim.read "${NAME}"
uaac --trace member add scim.write "${NAME}"
uaac --trace member add doppler.firehose "${NAME}"

echo "Removing user ${NAME}"
uaac --trace user delete "${NAME}"
```

**Comment by combor** [comment]
Now I'm thinking that I shall create new issue in https://github.com/cloudfoundry/cf-uaac repo.

**Comment by plfx** [comment]
@combor Sorry, I didn't realize you were using the UAAC. Yes, please open an issue on the CF-UAAC repo. In the meantime, as a workaround you can use the `uaac curl` command in your script to use the endpoints I described. Example:
```
uaac curl '/Groups/group-uuid/members' -XPOST -H'Content-Type: application/json' -d'{"origin":"uaa","type":"USER","value":"user-uuid"}'
```
Replace `user-uuid` and `group-uuid` with the appropriate values. This should not suffer from the issue you're experiencing now.

**Comment by mtekel** [comment]
Created new issue in uaac: https://github.com/cloudfoundry/cf-uaac/issues/25
